### PR TITLE
Adapter aliases: moved adapter related configs to bidder.yaml files

### PR DESCRIPTION
**Pull request by VeronikaSolovei9**
- all adapter related configs are moved to static/bidder-info/{bidder}.yaml files.
- Apapter struct now carries only adapter needed data
- all adapter related configs are removed from config.go
- bidder list now loads from the file system

Docs update PR: https://github.com/prebid/prebid.github.io/pull/4008

### CODE REVIEW

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [inline config/bidderinfo.go:365]: Will add unit tests for this function when logic is confirmed

**Comment by AlexBVolcy** [comment]
Also just wanted to comment highlighting the merge conflicts with bidderinfo., bidderinfo_test.go, adapter_util_test.go

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go:459]: Right, this is a part of merging function, will add it in next push

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go:498]: Will do!

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go:224]: This is moved from another place, I didn't change content. Renamed for future convenience.

File: config/bidderinfo.go
Comment (hhhjort) [inline config/bidderinfo.go:191]: Why do we have `LoadBidderInfoFromDisk(path)` when `loadBidderInfo(path)` is also necessarily from disk? I could see the utility if `loadBidderInfo()` was more general in some sense, but I don't see what we gain from it in this instance.

File: config/config.go
Comment (hhhjort) [inline config/config.go:911]: without any defaults, how will viper know to look for adapter config overrides in the environment variables. Viper does not use reflect to interrogate the config struct for all possible fields that might exist. I do think viper has another function other than `SetDefault()` that will clue it in to a field's existence without actually setting a default value.

File: config/config_test.go
Comment (hhhjort) [inline config/config_test.go:1382]: Why is this test being removed? Is it intended that it should no longer be possible to set usersync parameters from environment variables?

Comment (VeronikaSolovei9) [reply config/config_test.go:1382]: There will be no adapter related config in environment variables. All bidder related data is loading from {bidder}.yaml files or pbs.json or pbs.yaml files. 
`cfg.Adapters` were removed from Configuration struct. 

`assert.Nil(t, cfg.Adapters["appnexus"].Syncer.IFrame)` - is not a valid structure anymore.

File: config/config.go
Comment (VeronikaSolovei9) [reply config/config.go:911]: This also relates to Configuration -> map[string]Adapter which was removed. There is no way to set adapterDefaults to adapters. Unless i understand something wrong. We need to discuss it with the team.

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go:191]: Per my understanding it was done specifically for unit tests to mock data reading from file system. Implementations for "infoReader" are `infoReaderFromDisk` and `fakeInfoReader`. 

Do we want to support other types of data loading besides file system?

File: config/config_test.go
Comment (hhhjort) [reply config/config_test.go:1382]: It is common for kubernettes deployments to use environment variables for configuration rather than config files, particularly when you have deployments in multiple datacenters that have datacenter specific parameters (such as the bidder endpoint). Is this breaking change already documented in an issue? Otherwise we should certainly document it as such and give hosts ample opportunity to rework their setups to continue to work properly when this feature is removed.

File: config/config.go
Comment (hhhjort) [reply config/config.go:911]: See my comments below on environment variable config support being removed.

File: config/bidderinfo.go
Comment (hhhjort) [reply config/bidderinfo.go:191]: My confusion rests more in that if `LoadBidderInfoFromDisk()` is an `infoReader` interface, then `loadBidderInfo()` is as well, as they have the same signature. Normally when I see a pattern like this, it is of the form:
```f1(A) X {
  return f2(A,B)
  }```
  Where B is some constant, and `f2` is some more general function than `f1` is supposed to be. Such as if `loadBidderInfo()` is intended to support other sources than disk some day, but where `LoadBidderInforFromDisk()` is called, will only ever be loading from disk. Using the pattern you have used here would allow for expanding `loadBidderInfo()` in the future without having to update the parts of code that currently call it. Since I am not aware of any plans to support other data sources, and `loadBidderInfo` is also an implementation of `infoReader`, I still do not understand why both functions exist like this.

File: config/bidderinfo.go
Comment (AlexBVolcy) [inline config/bidderinfo.go:195]: Is it guaranteed that there will be data present at index 0 here? Just want to check

File: config/bidderinfo.go
Comment (AlexBVolcy) [inline config/bidderinfo.go:336]: Could this variable `v` be more descriptive? I understand if you want to leave it the way it is, but I tend to like a little more from variable names

**Review by AlexBVolcy — COMMENTED** [review]
Very minor comments

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go]: good point. Added check if file is a yaml file and array len.

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go]: Renamed

File: config/config.go
Comment (hhhjort) [inline config/config.go:705]: Would it make sense to insert the `bidderInfos` loaded from disk into the `config` struct before the viper load step, so viper does the merge automatically? And then perhaps a validate infos step to handle the extra logic in the Overrides function.

Comment (VeronikaSolovei9) [reply config/config.go:705]: I tried it and found interesting behavior. In `setBidderDefaults` bidder infos in viper initialized with empty values. If any field for adapter was set through env variables or pbs.json or pbs.yaml file - Viper sets this value to bidder info.
If I set bidder info from file system to config before unmarshalling - Viper replaces everything for bidders that were "extra" configured through env variables. These bidders only have data from env variables and nothing from file system. Bidders that were not configured through env variables have data from file system. 

Hope I understand you correctly, basically you suggest to do something like this: 
```
// New uses viper to get our server configurations.
func New(v *viper.Viper, bidderInfos BidderInfos) (*Configuration, error) {
	var c Configuration
	c.BidderInfos = bidderInfos ---> set loaded bidder info from filesystem here
	if err := v.Unmarshal(&c); err != nil {
		return nil, fmt.Errorf("viper failed to unmarshal app config: %v", err)
	}
      .....
```

Comment (VeronikaSolovei9) [reply config/config.go:705]: Could you please add more details about `validate infos step to handle the extra logic in the Overrides function`? Move bidder infos validation to `applyBidderInfoConfigOverrides`?

Comment (hhhjort) [reply config/config.go:705]: The above is what I was thinking. So you discovered that if, for example, a bidder's `endpoint` was set in an environment variable, viper would override the initial setting with that from the environment variable. However if it was set in `pbs.yaml`, viper would ignore it and keep what was originally set before the unmarshal? If this is true, there is some sort of inconsistency with viper. I don't think it will be worthwhile debugging that inconsistency, my suggestion was mainly in hopes of keeping the code simpler and easier to read, there should be no real functional advantage in doing it this way.

**Comment by AlexBVolcy** [comment]
This looks good! I want to confirm that exchange about bidderInfos and viper is resolved before approving!

File: config/config.go
Comment (VeronikaSolovei9) [reply config/config.go:705]: Discussed this offline. I didn't find better and easier way to achieve same result using Viper functionality.

File: config/config_test.go
Comment (hhhjort) [reply config/config_test.go:1382]: Did you add back in tests for environment variables setting adapter parameters somewhere else?

Comment (VeronikaSolovei9) [reply config/config_test.go:1382]: Good point! Added and fixed mapstructures in bidder info and added remaining bidder info properties to `setBidderDefaults`

**Review by hhhjort — DISMISSED** [review]
LGTM

**Review by AlexBVolcy — APPROVED** [review]

File: config/bidderinfo.go
Comment (SyntaxNode) [inline config/bidderinfo.go:318]: Nitpick: There is no need to include the wording "should be an array". Of course it's an array. Consider changing to an error along the line of "at least one meda type needs to be specified".

File: config/bidderinfo.go
Comment (SyntaxNode) [inline config/bidderinfo.go:344]: This is no longer necessary. We can let Viper take care of this work. You need to only give Viper the bidder infos before calling unmarshal and it will take care of the overrides:

```
func New(v *viper.Viper, bidderInfos BidderInfos) (*Configuration, error) {
	var c Configuration
	c.BidderInfos = bidderInfos

	if err := v.Unmarshal(&c); err != nil {
		return nil, fmt.Errorf("viper failed to unmarshal app config: %v", err)
	}
	
	...
```

This breaks is the legacy user sync url, which we can either choose to remove as part of this work or we alert `applyBidderInfoConfigOverrides` to only apply the legacy user sync url override. 

This also removes our ability to disallow certain parts of the config to be changed, like the maintainer or media types. The later can cause a problem if it's less restrictive.

I think we can either:
1. Let Viper handle everything and remove the override code. This allows hosts to "shoot themselves in the foot", but there's nothing stop them from editing the bidder info files directly?
2. Keep control of overrides and create a new BidderInfoOverride with less fields and leave the override code.

Thoughts?

File: config/config.go
Comment (SyntaxNode) [inline config/config.go:1143]: There be in danger in allowing unrestricted overrides of `capabilities.app.mediaTypes` and `capabilities.site.mediaTypes`.

File: config/config_test.go
Comment (SyntaxNode) [inline config/config_test.go:882]: IMHO we don't need to test all of the env variables, just a few is fine to provide viper binding works ok. This seems a little too verbose IMHO.

File: config/test/bidder-info-invalid/someBidder-invalid.yaml
Comment (SyntaxNode) [inline config/test/bidder-info-invalid/someBidder-invalid.yaml:3]: Why do we need do add this? Unit tests should cover this already.

File: endpoints/events/vtrack.go
Comment (SyntaxNode) [inline endpoints/events/vtrack.go:270]: I recommend adding a helper method to bidder info so we can call .IsEnabled() instead of negating the disabled field. The later is harder to read and easier to miss.

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go]: Modified

File: config/config.go
Comment (VeronikaSolovei9) [reply config/config.go:1141]: removed support of the capabilities override

File: config/config_test.go
Comment (VeronikaSolovei9) [reply config/config_test.go:827]: Removed some of checks

File: config/test/bidder-info-invalid/someBidder-invalid.yaml
Comment (VeronikaSolovei9) [reply config/test/bidder-info-invalid/someBidder-invalid.yaml]: removed. Just wanted to test it with actual yaml file.

File: config/bidderinfo.go
Comment (VeronikaSolovei9) [reply config/bidderinfo.go:348]: This is what you showed me on our very first discussion about the approach to support env variables. 
We also had a discussion about it with Hans: https://github.com/prebid/prebid-server/pull/2353#discussion_r963124663

If I set `c.BidderInfos = bidderInfos` and have env variable for at least one field for a bidder (for example endpoint) - it overrides the entire object for this bidder and sets "endpoint" only. Other data from yaml file is not present for this bidder after config umnarshalling. 

I think we need to control overrides. I removed `capabilities` from overrides, we can discuss what other fields should be removed. But I also think this is not obvious why some of the fields can be overridden and some not. Please advice.

File: config/bidderinfo.go
Comment (hhhjort) [inline config/bidderinfo.go:322]: media has a typo 'meda'

**Review by AlexBVolcy — DISMISSED** [review]

**Review by AlexBVolcy — APPROVED** [review]
LGTM

**Review by hhhjort — APPROVED** [review]

**Pull request merged**
